### PR TITLE
FV: Recovery can be always finalised

### LINKED ISSUE #11: FV: Finalization is always possible with suitable conditions
## Intention

> Can always finalize recovery once the delay period is over and if the owner has not cancelled.

## Description

The last step is to finalize the recovery process, at which point the old owners are removed and the new owners are added. Once we reach this state, unless an owner interferes, the finalization process should always be executed successfully.

## Expected Outcome

Rule to verify that a finalization step always goes through successfully if:
- the delay period is complete
- the owner does not interfere

### PR DESCRIPTION
**Pull request by akshay-ap**
Fixes #11 

Changes in PR:
- Adds a rule that verifies that recovery is always possible given certain conditions are met. These conditions are mentioned in the spec comments. These additional conditions on top of the 2 conditions given in the issue are required because proves reports counter example without them.

- This rule is different from `finalizeRecovery`. The new rule here verifies that recovery should always work when certain conditions are met.

### CODE REVIEW

File: certora/specs/SocialRecoveryModule.spec [hunk 1]
```diff
@@ -466,3 +466,27 @@ rule invalidatingNonceInRecovery(env e, address guardian, address[] newOwners, u
...
+rule finalizeRecoveryAlwaysPossible(env e) {
+    uint64 executeAfter = currentContract.recoveryRequests[safeContract].executeAfter;
+    require currentContract.walletsNonces[safeContract] > 0;
+    require forall uint256 i. i < currentContract.recoveryRequests[safeContract].newOwners.length =>
+            currentContract.recoveryRequests[safeContract].newOwners[i] != SENTINEL() &&
+            currentContract.entries[safeContract].guardians[currentContract.recoveryRequests[safeContract].newOwners[i]] == 0;
+
+    require safeContract.getOwners().length > 0;
+    require e.msg.value == 0;
+    require require_uint64(e.block.timestamp) > executeAfter;
```
Comment (nlordell) [inline certora/specs/SocialRecoveryModule.spec]:
```suggestion
    require require_uint64(e.block.timestamp) >= executeAfter;
```

File: certora/specs/SocialRecoveryModule.spec [hunk 2]
```diff
@@ -466,3 +466,27 @@ rule invalidatingNonceInRecovery(env e, address guardian, address[] newOwners, u
...
+    require executeAfter > 0;
```
Comment (nlordell) [inline certora/specs/SocialRecoveryModule.spec:486]: nit: would be nice to explain why this `require` is needed:

```suggestion
    // `executeAfter == 0` indicates the absence of a recovery.
    require executeAfter != 0;
```

File: certora/specs/SocialRecoveryModule.spec [hunk 3]
```diff
@@ -466,3 +466,27 @@ rule invalidatingNonceInRecovery(env e, address guardian, address[] newOwners, u
     bool isReverted = lastReverted;
     assert success => isReverted;
 }
+
+// This rule verifies that Recovery can be finalized after the delay period.
+// This rule requires other conditions to be met as well:
+// - The recovery request should be initiated.
```
Comment (nlordell) [inline certora/specs/SocialRecoveryModule.spec]:
```suggestion
// - The recovery request should be initiated (i.e. `executeAfter != 0` and `walletsNonce[safeContract] > 0`).
```

File: certora/specs/SocialRecoveryModule.spec [hunk 4]
```diff
@@ -466,3 +466,27 @@ rule invalidatingNonceInRecovery(env e, address guardian, address[] newOwners, u
...
+// - No ether should be sent with the transaction.
+// - New owner should not be a guardian.
+// - Existing Safe owner count should be more than zero.
+rule finalizeRecoveryAlwaysPossible(env e) {
+    uint64 executeAfter = currentContract.recoveryRequests[safeContract].executeAfter;
+    require currentContract.walletsNonces[safeContract] > 0;
```
Comment (nlordell) [inline certora/specs/SocialRecoveryModule.spec:496]: Why is this needed? Is there special meaning to `currentContract.walletsNonces[safeContract] == 0`?

**Edit**: `walletsNonces` can't be 0 for an initialized recovery. Suggested an edit to the comments so this is a bit clearer.

Comment (akshay-ap) [reply certora/specs/SocialRecoveryModule.spec:496]: This `require` is needed because prover reports a counter example where recovery reverts when the nonce is zero.

Comment (nlordell) [reply certora/specs/SocialRecoveryModule.spec:496]: > prover reports a counter example where recovery reverts when the nonce is zero.


I meant more as in "why does the prover find a counter example?". I found it in the contracts, and it is because once the recovery is initiated, the nonce is incremented, so it is impossible to have `walletsNonce[safe] == 0` before a recovery has initiated. This wasn't super clear to me without reading the code (and why I suggested the edits to the documentation to clarify this bit).

File: certora/specs/SocialRecoveryModule.spec [hunk 5]
```diff
@@ -484,3 +484,27 @@ rule doesNotAffectOtherAccount(env e, method f, calldataarg args, address otherS
...
+
+    require safeContract.getOwners().length > 0;
+    require e.msg.value == 0;
+    require require_uint64(e.block.timestamp) >= executeAfter;
+    require executeAfter > 0;
+
+    currentContract.finalizeRecovery@withrevert(e, safeContract);
+    bool isReverted = lastReverted;
+
+    assert !isReverted;
```
Comment (mmv08) [inline certora/specs/SocialRecoveryModule.spec]:
```suggestion
    assert !isReverted, "legitimate recovery finalization reverted";
```
nit

**Review by mmv08 — APPROVED** [review]
